### PR TITLE
Fix heatmap perf part 2

### PR DESCRIPTION
**Pull request by gregdingle**
# Description

This PR reduces the time to render the heatmap by about 40% on top of previous gains. In my test of a 49 sample project, time for the `samples_taxons` endpoint went down from 14s to 8s consistently. 

The improvement was achieved mostly by ranking and filtering in SQL instead of ruby. See `top_n_query`. 

Next, about 2 seconds was saved by converting two N+1 queries into a subquery to get the most recent pipeline run. 

Thirdly, the bottleneck SQL query was filtered more with `MINIMUM_ZSCORE_THRESHOLD` and `MINIMUM_READ_THRESHOLD`.

In the process, I refactored `HeatmapHelper` and `ReportHelper` a bunch. I moved all methods used only by `HeatmapHelper` into that class. There is still a lot of duplication of code in the two classes. I will leave that for another day. 

Other changes:
* I added a suite of unit tests to cover the public methods of `ReportHelper`. The tests cover different permutations of inputs but not all edge cases. 
* I removed the `aggregatescore` from the heatmap UI and code on recommendation by @cdebourcy .

# Notes

* **failed pipeline runs**: The original logic which I reproduced in SQL is to use the most recent run regardless of status. This doesn't seem right to me. 
* **missing pipeline runs**: I literally spent 3 hours debugging an issue which turned out to be caused by taxon counts referencing pipeline runs that did not exist. There appears to be 87 of such counts in prod. See https://modeanalytics.com/editor/chanzuckerberg/reports/46d9ac9db658 . It would be nice to enforce data integrity with either database constraints or garbage collecting scripts to avoid such head-scratching situations. 
* I **duplicated definitions** of user facing metrics such as zscore for performance reasons. This could easily lead to inconsistencies if not now then later which may be spotted by users. That said, there were **already** several duplicate definitions in the reports codebase, which is more troubling. 
* There are a bunch of **magic constants and hidden filtering** in the reports codebase without explanation. For example, phages are excluded by default without any indication in the UI.
* It seems like the **aggregate score** is no longer loved. If so we should kill it in the report page as well. 

# Test

`rails t` to cover basic param functionality

To cover metric definitions:

print the values of zscore and rpm, computed in ruby and in SQL
see the values are the same

To cover correct top N behavior: 

Load a large heatmap in dev branch
Load a large heatmap in master branch
See that the values on hover are the same
See that the grid of colors is very similar, differing only in equally ranked samples or taxons
Try all different metrics, different taxons per sample

To cover metric definitions:

Load a large heatmap in dev branch
print the values of zscore and rpm, computed in ruby and in SQL
see the values are the same

### CODE REVIEW

File: app/controllers/visualizations_controller.rb [hunk 1]
```diff
@@ -126,9 +126,67 @@ def shorten_url
     }, status: :internal_server_error
   end
 
+  # START OF HEATMAP METHODS
+
+  def heatmap
```
Comment (gregdingle) [inline app/controllers/visualizations_controller.rb]: I moved all action methods from the `HeatmapHelper` to the controller itself so that the helper could be a pure static class.

File: app/controllers/visualizations_controller.rb [hunk 2]
```diff
@@ -126,9 +126,67 @@ def shorten_url
...
+  end
+
   private
 
   def visualization_params
     params.permit(:domain, :type, :id, :url, :search, data: {})
   end
+
+  def samples_for_heatmap
+    current_power.samples
```
Comment (tfrcarvalho) [inline app/controllers/visualizations_controller.rb]: Since we are moving, this looks like a model kind of function.

File: app/helpers/heatmap_helper.rb [hunk 3]
```diff
@@ -1,61 +1,25 @@
+# This is a class of static helper methods for generating data for the heatmap
+# visualization. See HeatmapHelperTest.
+# See selectedOptions in SamplesHeatmapView for client-side defaults, and
+# heatmap action in VisualizationsController.
 module HeatmapHelper
   DEFAULT_MAX_NUM_TAXONS = 30
+  # Zscore is best for heatmaps because it weighs the frequency against the background
+  DEFAULT_TAXON_SORT_PARAM = 'highest_nt_zscore'.freeze
```
Comment (tfrcarvalho) [inline app/helpers/heatmap_helper.rb]: Where did you get this information?
Z-score does not differentiate much by itself.

Comment (tfrcarvalho) [inline app/helpers/heatmap_helper.rb]: Is this new?

File: app/helpers/heatmap_helper.rb [hunk 4]
```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
...
+        taxon_counts.percent_concordant  AS  percentconcordant
+      FROM taxon_counts
+      LEFT OUTER JOIN taxon_summaries ON
+        #{background_id.to_i}   = taxon_summaries.background_id   AND
+        taxon_counts.count_type = taxon_summaries.count_type      AND
+        taxon_counts.tax_level  = taxon_summaries.tax_level       AND
+        taxon_counts.tax_id     = taxon_summaries.tax_id
+      WHERE
+        pipeline_run_id IN (
+          SELECT MAX(id)
```
Comment (tfrcarvalho) [inline app/helpers/heatmap_helper.rb]: Max id is not necessarily the last run... This is something we need to fix in the entire app...eventually.

File: app/helpers/heatmap_helper.rb [hunk 5]
```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
...
+          a.*
+        FROM (
+          #{query}
+        ) a
+        ORDER BY
+          pipeline_run_id,
+          count_type = '#{sort[:count_type]}' DESC,
+          #{sort[:metric]} #{sort[:direction] == 'highest' ? 'DESC' : 'ASC'}
+      ) b
+      -- Overfetch by a factor of 4 to allow for post-SQL threshold filters
```
Comment (cdebourcy) [inline app/helpers/heatmap_helper.rb]: Would appreciate more explanation here. Which filters are they? Why are they applied post-SQL? Is there a way to apply them in SQL? Is there a risk that the results are going to be "inexact"? Is there a way to make them exact?

File: app/controllers/visualizations_controller.rb (hunk 2 above)
Comment (gregdingle) [reply app/controllers/visualizations_controller.rb]: isn't `current_power` required to be in a controller because it uses session state?

File: app/helpers/heatmap_helper.rb (hunk 3 above)
Comment (gregdingle) [reply app/helpers/heatmap_helper.rb]: yes, this was my reasoning and it could be wrong -- thanks for asking. @cdebourcy told me to remove the aggregate score but he didn't say what should replace it as a default.

File: app/helpers/heatmap_helper.rb [hunk 6]
```diff
@@ -1,61 +1,25 @@
...
+  # Zscore is best for heatmaps because it weighs the frequency against the background
+  DEFAULT_TAXON_SORT_PARAM = 'highest_nt_zscore'.freeze
+  READ_SPECIFICITY = true
+  MINIMUM_READ_THRESHOLD = 5
+  # Note: this is activated from the heatmap page by selecting "Viruses -
+  # Phages". The default categories are all BUT phages, though the UI does not
+  # indicate this.
+  INCLUDE_PHAGE = false
 
-  def heatmap
```
Comment (gregdingle) [inline app/helpers/heatmap_helper.rb]: moved to controller

File: app/helpers/heatmap_helper.rb [hunk 7]
```diff
@@ -1,61 +1,25 @@
...
-      allowedFeatures: current_user.allowed_feature_list
-    }
-  end
-
-  def download_heatmap
-    @sample_taxons_dict = sample_taxons_dict(params)
-    output_csv = generate_heatmap_csv(@sample_taxons_dict)
-    send_data output_csv, filename: 'heatmap.csv'
-  end
+  MINIMUM_ZSCORE_THRESHOLD = 1.7
```
Comment (gregdingle) [inline app/helpers/heatmap_helper.rb]: moved from report helper

File: app/helpers/heatmap_helper.rb [hunk 8]
```diff
@@ -1,61 +1,25 @@
...
-    @sample_taxons_dict = sample_taxons_dict(params)
-    render json: @sample_taxons_dict
-  end
+  def self.sample_taxons_dict(params, samples)
+    return {} if samples.empty?
 
-  def sample_taxons_dict(params)
-    sample_ids = (params[:sampleIds] || []).map(&:to_i)
     num_results = params[:taxonsPerSample] ? params[:taxonsPerSample].to_i : DEFAULT_MAX_NUM_TAXONS
+    min_reads = params[:minReads] ? params[:minReads].to_i : MINIMUM_READ_THRESHOLD
```
Comment (gregdingle) [inline app/helpers/heatmap_helper.rb]: this is now dynamic just so I could have fewer fixtures in the unit test

File: app/helpers/heatmap_helper.rb [hunk 9]
```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
                           JSON.parse(params[:thresholdFilters] || "[]")
                         end
     subcategories = if params[:subcategories] && params[:subcategories].respond_to?(:to_h)
-                      params[:subcategories].permit!.to_h
+                      params[:subcategories].to_h
```
Comment (gregdingle) [inline app/helpers/heatmap_helper.rb]: to make it a static method... not sure if it's important to always `permit`

File: app/helpers/heatmap_helper.rb [hunk 10]
```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
...
+    HeatmapHelper.samples_taxons_details(
+      samples,
+      taxon_ids,
+      background_id,
+      species_selected,
+      threshold_filters
+    )
+  end
+
+  def self.top_taxons_details(
```
Comment (gregdingle) [inline app/helpers/heatmap_helper.rb]: moved from heatmap helper and functionally untouched internally

File: app/helpers/heatmap_helper.rb [hunk 11]
```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
...
+        taxon = if candidate_taxons[row["tax_id"]]
+                  candidate_taxons[row["tax_id"]]
+                else
+                  { "tax_id" => row["tax_id"], "samples" => {} }
+                end
+        taxon["max_aggregate_score"] = row[sort[:count_type]][sort[:metric]] if
+          taxon["max_aggregate_score"].to_f < row[sort[:count_type]][sort[:metric]].to_f
+        taxon["samples"][sample_id] = [count, row["tax_level"], row["NT"]["zscore"], row["NR"]["zscore"]]
+        candidate_taxons[row["tax_id"]] = taxon
+        break if count >= num_results
```
Comment (gregdingle) [inline app/helpers/heatmap_helper.rb]: the original sorting and ranking behavior is preserved and applied on top of the SQL sort and rank so that results are guaranteed to be the same as before

File: app/helpers/heatmap_helper.rb [hunk 12]
```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
...
+        count += 1
+      end
+    end
+
+    candidate_taxons.values.sort_by { |taxon| -1.0 * taxon["max_aggregate_score"].to_f }
+  end
+
+  def self.fetch_top_taxons(
```
Comment (gregdingle) [inline app/helpers/heatmap_helper.rb]: moved from report helper... original SQL query is slightly modified and wrapped in outer SQL query that sorts and ranks emulating logic that is in ruby... this way we avoid processing 100k+ rows in ruby

File: app/helpers/heatmap_helper.rb (hunk 5 above)
Comment (gregdingle) [reply app/helpers/heatmap_helper.rb]: will discuss in person

File: app/helpers/heatmap_helper.rb (hunk 4 above)
Comment (gregdingle) [reply app/helpers/heatmap_helper.rb]: good point ... should disucss in person

File: app/helpers/heatmap_helper.rb (hunk 5 above)
Comment (gregdingle) [reply app/helpers/heatmap_helper.rb]: After discussion, I decided to move the logic of the threshold filters into the SQL.

File: app/helpers/heatmap_helper.rb [hunk 13]
```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
...
+        row["zscore"] = z_max if row["zscore"] > z_max &&
+                                 row["zscore"] != ReportHelper::ZSCORE_WHEN_ABSENT_FROM_BACKGROUND
+        row["zscore"] = z_min if row["zscore"] < z_min
+        result_hash[pipeline_run_id]["taxon_counts"] << row
+      end
+    end
+    result_hash
+  end
+
+  def self.samples_taxons_details(
```
Comment (gregdingle) [inline app/helpers/heatmap_helper.rb]: this method and the rest below were moved with no change in logic from report helper

File: app/controllers/visualizations_controller.rb (hunk 2 above)
Comment (tfrcarvalho) [reply app/controllers/visualizations_controller.rb]: Yeah, nevermind, I guess the key here for samples is the includes statement which is heatmap specific. 
I would just see the rest as something like `current_power.visualization.authorized_samples` which you would define through power controller. 

In any case, disregard this comment on the context of this PR.

File: app/helpers/heatmap_helper.rb (hunk 3 above)
Comment (tfrcarvalho) [reply app/helpers/heatmap_helper.rb]: I belive R and/or RPMs are the best since the absolute value of the difference is also relevant. The problem with Z is that it relies on an (often) invalid distribution and loses precision quickly (lots of entries close max value 100)

File: app/helpers/heatmap_helper.rb (hunk 8 above)
Comment (tfrcarvalho) [reply app/helpers/heatmap_helper.rb]: One of the reasons we are moving to factories... need to get my PR out soon.

File: app/helpers/heatmap_helper.rb [hunk 14]
```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
...
+    count_type = sort[:count_type]
+    metric = sort[:metric]
+    candidate_taxons = {}
+    results_by_pr.each do |_pr_id, res|
+      pr = res["pr"]
+      taxon_counts = res["taxon_counts"]
+      sample_id = pr.sample_id
+
+      tax_2d = ReportHelper.taxon_counts_cleanup(taxon_counts)
+      HeatmapHelper.only_species_or_genus_counts!(tax_2d, species_selected)
```
Comment (tfrcarvalho) [inline app/helpers/heatmap_helper.rb]: Cant this be a filter on SQL too?

**Comment by gregdingle** [comment]
@cdebourcy @tfrcarvalho 

I spent another 1.5 hours today trying to convert the custom threshold filters and all the remaining ruby filtering to SQL. However, I gave up because it was too hard to get exactly the same results in all cases. The key difference is that the ruby filters are applied at the level of taxons but the inner SQL query is applied on counts. 

The good news is that custom filtering is a small percent of queries and we could easily route to the old code path, which I implemented in the lastest commit, so we get nearly all the performance benefit at the cost of code complexity. 

Before working any further on heatmaps, I believe we need to refactor the code a lot more and write tests for each part, assuming we want to maintain the current behavior exactly.

**Comment by tfrcarvalho** [comment]
I am ok with moving on with this change. I understand some refactors are too long for the current scope.
